packages/fee-estimation/src/estimateFees.ts

```diff
@@ -0,0 +1,286 @@
+  validateFetch()
+
+  if ('chainId' in options && options.chainId) {
+    const viemChain = Object.values(chains)?.find((chain) => chain.id === options.chainId)
+    const rpcUrls = options.rpcUrl
+      ? { default: { http: [options.rpcUrl] }, public: { http: [options.rpcUrl] } }
+      : viemChain?.rpcUrls
+    if (!rpcUrls) {
+      throw new Error(
+        `No rpcUrls found for chainId ${options.chainId}.  Please explicitly provide one`
+      )
+    }
+    return createPublicClient({
+      chain: {
+        id: options.chainId,
+        name: viemChain?.name ?? 'op-chain',
+        nativeCurrency:
+          options.nativeCurrency ??
+          viemChain?.nativeCurrency ??
+          chains.optimism.nativeCurrency,
+        network: viemChain?.network ?? 'Unknown OP Chain',
+        rpcUrls,
+        explorers:
+          (viemChain as typeof chains.optimism)?.blockExplorers ??
+          chains.optimism.blockExplorers,
+      },
+      transport: http(options.rpcUrl ?? chains[options.chainId].rpcUrls.public.http[0]),
+    })
+  }
+  return options as PublicClient
+}
+
+/**
+ * Get gas price Oracle contract
+ */
+export const getGasPriceOracleContract = (params: ClientOptions) => {
+  return getContract({
+    address: gasPriceOracleAddress['420'],
+    abi: gasPriceOracleABI,
+    publicClient: getL2Client(params),
+  })
+}
+
+/**
+ * Returns the base fee
+ * @returns {Promise<bigint>} - The base fee
+ * @example
+ * const baseFeeValue = await baseFee(params);
+ */
+export const baseFee = async (
+  { client, ...params }: GasPriceOracleOptions
+): Promise<bigint> => {
+  const contract = getGasPriceOracleContract(client)
+  return contract.read.baseFee(params)
+}
+
+/**
+ * Returns the decimals used in the scalar
+ * @example
+ * const decimalsValue = await decimals(params);
+ */
+export const decimals = async (
+  { client, ...params }: GasPriceOracleOptions
+): Promise<bigint> => {
+  const contract = getGasPriceOracleContract(client)
+  return contract.read.decimals(params)
+}
+
+/**
+ * Returns the gas price
+ * @example
+ * const gasPriceValue = await gasPrice(params);
+ */
+export const gasPrice = async (
+  { client, ...params }: GasPriceOracleOptions
+): Promise<bigint> => {
+  const contract = getGasPriceOracleContract(client)
+  return contract.read.gasPrice(params)
+}
+
+/**
+ * Computes the L1 portion of the fee based on the size of the rlp encoded input
+ * transaction, the current L1 base fee, and the various dynamic parameters.
+ * @example
+ * const L1FeeValue = await getL1Fee(data, params);
+ */
+export const getL1Fee = async (
+  data: Bytes,
+  { client, ...params }: GasPriceOracleOptions
+): Promise<bigint> => {
+  const contract = getGasPriceOracleContract(client)
+  return contract.read.getL1Fee([data], params)
+}
+
+/**
+ * Returns the L1 gas used
+ * @example
+ */
+export const getL1GasUsed = async (
+  /**
+   * The transaction call data as a 0x-prefixed hex string
```

Comment:
This is meant to be the RLP-serialized transaction and NOT the transaction calldata, right?

Reply:
Correct good catch
```suggestion
   * Unsigned fully RLP-encoded transaction to get the L1 fee for
```

Reply:
@jxom Do I need to use https://viem.sh/docs/utilities/toRlp.html#torlp after calling `encodeFunctiondata`?  I'm reexporting encodeFunctionData as a helper to help create this hex string

Reply:
No, you’d need to call `serializeTransaction`. 

I’d highly recommend for the SDK to do this, and accept the transaction request attributes as params instead of a hex value.

Reply:
`accept the transaction request attributes as params instead of a hex value.`
Now that you say it it seems obvious definitely will do